core/store/src/flat/manager.rs

```diff
@@ -0,0 +1,136 @@
+    /// This may cause some overhead because the data like shards that the node is processing for
+    /// this epoch can share the same `head` and `tail`, similar for shards for the next epoch,
+    /// but such overhead is negligible comparing the delta sizes, so we think it's ok.
+    flat_storages: Mutex<HashMap<ShardId, FlatStorage>>,
+}
+
+impl FlatStorageManager {
+    pub fn new(store: Store) -> Self {
+        Self(Arc::new(FlatStorageManagerInner { store, flat_storages: Default::default() }))
+    }
+
+    /// When a node starts from an empty database, this function must be called to ensure
+    /// information such as flat head is set up correctly in the database.
+    /// Note that this function is different from `add_flat_storage_for_shard`,
+    /// it must be called before `add_flat_storage_for_shard` if the node starts from
+    /// an empty database.
+    pub fn set_flat_storage_for_genesis(
+        &self,
+        store_update: &mut StoreUpdate,
+        shard_id: ShardId,
+        genesis_block: &CryptoHash,
+    ) {
+        let flat_storages = self.0.flat_storages.lock().expect(POISONED_LOCK_ERR);
+        assert!(!flat_storages.contains_key(&shard_id));
+        store_helper::set_flat_head(store_update, shard_id, genesis_block);
+    }
+
+    /// Add a flat storage state for shard `shard_id`. The function also checks that
+    /// the shard's flat storage state hasn't been set before, otherwise it panics.
+    /// TODO (#7327): this behavior may change when we implement support for state sync
+    /// and resharding.
+    pub fn add_flat_storage_for_shard(&self, shard_id: ShardId, flat_storage: FlatStorage) {
+        let mut flat_storages = self.0.flat_storages.lock().expect(POISONED_LOCK_ERR);
+        let original_value = flat_storages.insert(shard_id, flat_storage);
+        // TODO (#7327): maybe we should propagate the error instead of assert here
+        // assert is fine now because this function is only called at construction time, but we
+        // will need to be more careful when we want to implement flat storage for resharding
+        assert!(original_value.is_none());
+    }
+
+    /// Creates `FlatStorageChunkView` to access state for `shard_id` and block `block_hash`. Note that
+    /// the state includes changes by the block `block_hash`.
+    /// `block_hash`: only create FlatStorageChunkView if it is not None. This is a hack we have temporarily
+    ///               to not introduce too many changes in the trie interface.
+    /// `is_view`: whether this flat state is used for view client. We use a separate set of caches
+    ///            for flat state for client vs view client. For now, we don't support flat state
+    ///            for view client, so we simply return None if `is_view` is True.
+    /// TODO (#7327): take block_hash as CryptoHash instead of Option<CryptoHash>
+    /// TODO (#7327): implement support for view_client
+    pub fn new_flat_state_for_shard(
```

Comment:
```suggestion
    pub fn get_chunk_view(
```

Makes sense to simplify naming here as well.
I wonder how we should name variables instead of `flat_state`. Maybe `flat_storage_chunk_view` or just `chunk_view` if context is obvious.

Reply:
good point, done: https://github.com/near/nearcore/pull/8664/commits/c8a1383702d0c90507384b8b32c807a1ff851e88